### ISSUE TITLE
devprod: consider `pre-commit`

### ISSUE THREAD

**Issue by austin3dickey** [conbench/conbench#768]
`pre-commit` has been really handy for me in the past to ensure my code is linted before committing. I was recently iterating quickly on https://github.com/conbench/conbench/pull/766 and kept pushing unlinted code. It would be really quick to set up (10 minutes?) and definitely save me some time in the long run.

I think I'd want to peg specific hook revisions in the `.pre-commit-config.yaml` and then run `pre-commit run --all-files` during the linting step in CI so that we always have version compatibility between local and CI.

But if this proposal is controversial, I'm happy to forget about it. 😉

**Comment by alistaire47** [comment]
I've occasionally found it makes my process slow or awkward and so turned it off and did it manually. (Well, `black` on save via vscode.) But in theory, if it's snappy, I like it for preventing all my `lint` and `isort` commits

**Comment by austin3dickey** [comment]
+1 for "only if it's fast." In this repo, it's pretty fast if you've run `mypy` before and have built up a `.mypy_cache/`.

**Comment by jgehrcke** [comment]
Pre-commit logic _can_ of course be used by individual developers if they choose to do so. But I am not really in favor of tracking that logic in the code repository, or suggesting to new developers that they "should" run certain pre-commit logic.

I have tried often, and always fell back to not doing that anymore.

Constraints that we care about must be implemented in CI, and I have seen that sometimes that idea gets diluted when building constraints into pre-commit logic that then is otherwise _not_ covered by CI. Split brain. Relying in CI on only the pre-commit approach (to keep things "in sync") is IMO too limiting.

Also; pre-commit things have never been fast enough for me, especially when doing larger rebases. I rebase a lot.

During rebase and potentially addressing merge commits I don't want isort&Co to mutate files between commits.

I do format-on-save etc in VSCode and otherwise try to display as many warnings/errors as possible in my editor. If I miss some things that are later caught by CI that's OK given the idea that the CI feedback arrives quickly. We have worked towards that!

I find that I often forget to run `make lint` before pushing /opening an PR; and that's where I can for myself save a minute (to catch an error a little earlier).
